src/NATS.Client.Core/Commands/ProtocolWriter.cs

```diff
@@ -48,70 +48,131 @@ public void WritePong()
         {
-            _writer.WriteNumber(payload.Length);
+            span[0] = (byte)'P';
+            span[1] = (byte)'U';
+            span[2] = (byte)'B';
+            span[3] = (byte)' ';
+            span = span[4..];
         }
         else
         {
-            var headersLength = _bufferHeaders.WrittenSpan.Length;
-            _writer.WriteNumber(CommandConstants.NatsHeaders10NewLine.Length + headersLength);
-            _writer.WriteSpace();
-            var total = CommandConstants.NatsHeaders10NewLine.Length + headersLength + payload.Length;
-            _writer.WriteNumber(total);
+            span[0] = (byte)'H';
+            span[1] = (byte)'P';
+            span[2] = (byte)'U';
+            span[3] = (byte)'B';
+            span[4] = (byte)' ';
+            span = span[5..];
         }
 
-        // End of message first line
-        _writer.WriteNewLine();
+        var written = _subjectEncoding.GetBytes(subject, span);
+        span[written] = (byte)' ';
+        span = span[(written + 1)..];
 
-        if (headers != null)
+        if (replyTo != null)
         {
-            _writer.WriteSpan(CommandConstants.NatsHeaders10NewLine);
-            _writer.WriteSpan(_bufferHeaders.WrittenSpan);
+            written = _subjectEncoding.GetBytes(replyTo, span);
+            span[written] = (byte)' ';
+            span = span[(written + 1)..];
         }
 
-        if (payload.Length != 0)
+        Span<byte> lenSpan;
+        if (headers == null)
         {
-            _writer.WriteSequence(payload);
+            // len =            payload len
+            lenSpan = span[..MaxIntStringLength];
+            span = span[lenSpan.Length..];
+        }
+        else
+        {
+            // len =          header len         +' '+ payload len
+            lenSpan = span[..(MaxIntStringLength + 1 + MaxIntStringLength)];
+            span = span[lenSpan.Length..];
         }
 
-        _writer.WriteNewLine();
-    }
+        span[0] = (byte)'\r';
+        span[1] = (byte)'\n';
+        _writer.Advance(ctrlLen);
 
-    public void WritePublish<T>(string subject, string? replyTo, NatsHeaders? headers, T? value, INatsSerialize<T> serializer)
-    {
-        _bufferPayload.Reset();
+        var headersLength = 0L;
+        var totalLength = 0L;
+        if (headers != null)
+        {
+            headersLength = _headerWriter.Write(headers);
+            totalLength += headersLength;
+        }
 
         // Consider null as empty payload. This way we are able to transmit null values as sentinels.
         // Another point is serializer behaviour. For instance JSON serializer seems to serialize null
         // as a string "null", others might throw exception.
         if (value != null)
         {
-            serializer.Serialize(_bufferPayload, value);
+            var initialCount = _writer.UnflushedBytes;
+            serializer.Serialize(_writer, value);
+            totalLength += _writer.UnflushedBytes - initialCount;
```

Comment:
Pardon my complete ignorance here (haven't pulled the branch down) but is there any chance `UnflushedBytes` could be racy?

Reply:
I don't think so, since access to the writer is serialized.  And we ensure that the previous Flush has completed every time prior to writing a new comand